### ISSUE TITLE
Feature request: Make available input parameters in a :params map in request

### ISSUE THREAD

**Issue by avallark** [kumarshantanu/calfpath#5]
Great work with Calfpath! Thank you.

Could you please make available a :params map within request with all the parameters, rather than adding them the request map directly? This would help identify all the parameters within a request. Thus we wont have to send around the request object to all my functions that use them.

**Comment by kumarshantanu** [comment]
Hey @avallark, initially during `0.5.0` development I had implemented the path params as a separate map in the request, which unfortunately degraded routing performance and made extracting the path params from request quite verbose and repetitive. The `0.5.0` version we have now, brings in Ring-handler compatibility while maintaining similar convenience and performance as in `0.4.x`, e.g. see below:

```clojure
(defn handler
  "Handler for /prefix/:foo/and/:bar/only/"
  [{:keys [foo bar] :as request}]
  (biz-logic foo bar))
```

However, I can understand that having path params separately would be useful, e.g. transparent logging, instrumentation etc. I'll keep this issue open as a feature request to add a knob to put path params into a separate map.

**Comment by avallark** [comment]
I got the functionality I needed by using 

```clj
ring.middleware.params/wrap-params
```

Thank you for your answer.

**Comment by kumarshantanu** [comment]
@avallark `ring.middleware.params/wrap-params` is for query and form params. I thought you meant it for path params. Any way, I'll consider path-params under a specified key in a future release.

**Comment by avallark** [comment]
I initially meant query and form params. But now as i use the path params more often,  this would be a nice feature to have.

A problem/bug here is that the path params that i am getting is url encoded for e.g. if i pass "Item Costs" as a path param value, the path param value i get in my handler is "Items%20Costs". I currently have to do a ring.util.codec/url-decode on all my path variables in my handler functions for paths with a path variable.

Anyway out of this ?

**Comment by avallark** [comment]
Sorry accidenrally clicked close on my phone! Cont d... 

The path param i get is "Items%20Costs". I currently have to now do a ring.util.codec/url-decode on all my handler functions for paths with a path variable. 

Anyway out of this ?

**Comment by kumarshantanu** [comment]
This is fixed in the `v0.6.0` branch via https://github.com/kumarshantanu/calfpath/pull/7/commits/8183059302ed224afa8d01da792e1c75f852eafd

@avallark The url-encoding issue can be fixed using a params key (described in this issue) and a middleware to apply `url-decode` on the param values.
